### PR TITLE
feat: dynamic datafiles table

### PR DESCRIPTION
**Pull request by abdellah257**
## Description
adding the option to opt for a dynamic datafiles table.

image

## Motivation
Supports custom columns, and a dynamic fetching of datafiles per batch using `/api/v3/origdatablocks/fullquery` instead of storing all files in memory with a `/api/v3/datasets/{pid}/origdatablocks` request.

## Changes:
This PR include multiple changes:
- Implementation of state management for `files` with a `datasetId` filter to support per batch loading of files.
- A wrapper to chose from `static-datafiles` (old view) or `dynamic-datatfiles` supported in this PR.
- Configuration options `dynamicDatafilesViewEnabled` and `defaultDatafilesColumnsList` to configure the dynamic datafiles table.
- `pipe` option at table columns configuration with 3 pipes currently supported (`filesize`, `date`, `timeDuration`).

## Tests included
- [x] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 

## Documentation
- [ ] swagger documentation updated \[required\]
- [x] official documentation updated \[nice-to-have\]

### official documentation info
If you have updated the official documentation, please provide PR # and URL of the pages where the updates are included

## Summary by Sourcery

Introduce a configurable, server-side backed dynamic datafiles table for dataset details, including selection state management and integration with existing tables and actions.

New Features:
- Add a dynamic datafiles table component that fetches dataset files via fullquery with server-side pagination, sorting, and configurable columns.
- Introduce configuration options to enable the dynamic datafiles view and define default datafiles columns and actions, along with per-user saved table column settings.
- Add support for formatting table cell values via configurable pipes (date, filesize, timeDuration) and customizable empty values in the shared dynamic table.

Bug Fixes:
- Prevent errors in configurable actions when variables are undefined and avoid repeated disabled-state evaluation issues.

Enhancements:
- Extend files NgRx state, actions, selectors, and effects to handle dataset-scoped origdatablocks, counts, and multi-selection, including selection persistence across pages.
- Wire the datasets module into the files feature store/effects and route the dataset datafiles tab through a wrapper that chooses static or dynamic implementation based on config.
- Refine table and models typing around datafiles (DataFile/FileOrigdatablock) and align configurable actions and static datafiles components with these types.
- Improve default list settings documentation to describe the dynamic datafiles table configuration and user-specific column settings.

Documentation:
- Update contributor documentation to describe configuration and behavior of the dynamic datafiles table, including default columns and user overrides.

### CODE REVIEW

File: src/app/shared/pipes/time-duration.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    const minutes = Math.floor(seconds / 60);
+    const hours = Math.floor(minutes / 60);
+    const leftMinutes = minutes % 60;
+    const leftSeconds = Math.round(seconds % 60);
+
+    if (hours === 0) {
+      if (leftMinutes === 0) return `${leftSeconds}s`;
+      return `${leftMinutes}m ${leftSeconds}s`;
+    }
+    return `${hours}h ${leftMinutes}m ${leftSeconds}s`;
```
Comment (nitrosx) [inline src/app/shared/pipes/time-duration.pipe.ts:19]: What if the duration becomes days or longer?
Do you know if there is any library who does this conversion ?

**Review by nitrosx — COMMENTED** [review]
I would modify the code so there is only one "Datafiles" tab.
When dynamic datafiles are enabled, the table under Datafiles tab becomes dynamic and configurable.
I can see that at the end, we will use only the dynamic version with a locked configuration to show it as it is now.

Please address the comments.

It would be great if you could create a branch directly on the main repository.

**Comment by nitrosx** [comment]
# Code Review Report: dynamic-datafile-table Branch

---

## Overview

1. **Logic Correctness**: The changes introduce a dynamic datafile table component with server-side pagination, selection management, and configurable columns. The logic is generally sound with proper state management using NgRx.

2. **Edge Cases Handled**: 
   - Empty values in table columns are handled with `emptyValue` configuration
   - Non-visible columns are accounted for when dropping (commits 121641cf, 9bc2bb79)
   - Loading states are properly managed
   - Selection state is maintained across pagination
   - File size warnings for large files are implemented

3. **Code Summary**: The changes implement a new dynamic datafile table component that replaces the static datafile view when enabled. It features:
   - Server-side pagination and sorting
   - Configurable columns via admin settings
   - Row selection with state management
   - Integration with existing action system
   - New pipes for time duration formatting
   - Session expiration handling via auth interceptor

4. **Changes Assessment**: The changes make sense and provide a significant improvement over the static datafile view, offering better performance for large datasets and more configuration options.

5. **Unreachable Code**: No obvious unreachable code identified in the main feature implementation.

**Key Commits**:
- `8decc493`: Main feature - dynamic datafile component
- `121641cf`, `9bc2bb79`: Fix column dropping with non-visible columns
- `8383d427`: Remove horizontal overflow
- `06e91c07`: Add loading spinner
- `66ac733a`: Session expired interceptor

---
---
## Code Changes

### Major Additions

| File | Change Type | Description |
|------|-------------|-------------|
| `dynamic-datafiles.component.ts` | New | Main component for dynamic datafile table (474 lines) |
| `auth.interceptor.ts` | New | HTTP interceptor for session expiration handling |
| `time-duration.pipe.ts` | New | Pipe for formatting time durations |
| `datafiles-wrapper.component.ts` | New | Wrapper to toggle between static/dynamic views |
| `datafiles-wrapper.component.spec.ts` | New | Tests for wrapper component |

### State Management Enhancements

- **Actions**: Added new actions for dataset origdatablock fetching and selection management
- **Reducers**: Enhanced files reducer with selection state management
- **Selectors**: Added selectors for dataset-specific origdatablocks and selection state
- **Models**: Added `FileOrigdatablock`, `DataFile`, `FilesDatasetFilter` interfaces

### Table Component Enhancements

- **dynamic-mat-table.component.ts**: Added pipe support (`filesize`, `timeDuration`, `date`) in columns
- **Empty value handling**: Improved with configurable `emptyValue` property
- **Column drag-drop**: Fixed to account for non-visible columns

### Configuration

- **config.json**: Added `dynamicDatafilesViewEnabled` flag and `defaultDatafilesColumnsList`
- **app-config.service.ts**: Added new config interface properties
- **docs/contributors/default-list-settings.md**: Updated documentation

### Assessment

**Necessity**: The changes are necessary to provide a scalable, configurable datafile table that can handle large datasets efficiently. The static table approach doesn't scale well for datasets with many files.

**Improvement Opportunities**:
1. The `AuthInterceptor` has unused imports (`Subscription`, `selectProfile`) that should be removed
2. Some type casting (`as unknown as FileOrigdatablock[]`) could be improved with proper type definitions
3. The `processSelector` method in `configurable-action.component.ts` was refactored but could benefit from more comprehensive error handling
4. The time duration pipe could handle edge cases better (negative values, non-numeric inputs)

**Verdict**: The changes are well-structured and provide significant value. The implementation follows Angular best practices and integrates well with the existing codebase.

---
---
## Security Review

### Findings

1. **Potential Injection Vulnerabilities**: 
   - **NONE FOUND**: No direct use of `innerHTML` with user-controlled data without sanitization. The `hasFileAboveMaxSizeWarning()` method uses `innerHTML` but only with configured template strings that contain placeholders replaced by controlled values from the app config.

2. **Sensitive User Data Exposure**:
   - **NONE FOUND**: No new endpoints exposing sensitive data. The changes primarily deal with frontend state management and UI rendering.

3. **Insecure API Usage**:
   - **NONE FOUND**: API calls use the existing service layer. The auth interceptor properly handles 401 errors with session expiration.

4. **Authentication Bypass**:
   - **NONE FOUND**: The auth interceptor strengthens security by properly handling session expiration and redirecting to login.

**Commits of Note**:
- `66ac733a`: Added session expired interceptor which improves security by properly handling expired sessions

---
---
## Test Coverage

### Current Coverage

| Component | Test File | Coverage Assessment |
|-----------|-----------|---------------------|
| DatafilesWrapperComponent | ✅ Yes | Good - tests both dynamic and static view selection |
| DynamicDatafilesComponent | ❌ No | **MISSING** - No dedicated test file |
| AuthInterceptor | ❌ No | **MISSING** - No tests for new interceptor |
| TimeDurationPipe | ❌ No | **MISSING** - No tests for new pipe |
| Files effects/reducers | ✅ Partial | Existing tests updated but may not cover new actions |

### Improvement Suggestions

1. **Critical**: Create test file for `DynamicDatafilesComponent` covering:
   - Initialization and data loading
   - Pagination and sorting
   - Row selection
   - Column configuration

2. **High Priority**: Create test file for `AuthInterceptor` covering:
   - Session expired detection
   - Redirect behavior
   - Error propagation

3. **Medium Priority**: Create test file for `TimeDurationPipe` covering:
   - Various time durations (seconds, minutes, hours)
   - Edge cases (0, negative values)
   - Format consistency

4. **Medium Priority**: Enhance existing reducer/effect tests to cover:
   - New dataset origdatablock actions
   - Selection state management
   - Error handling paths

**Relevant Commits**: All feature commits should include corresponding test additions.

---
---
## Security Examples

This section is empty - no relevant security vulnerabilities were identified.

---
---
## Testing for Security Use Cases

### Test Cases for AuthInterceptor

```typescript
// Example test cases that should be implemented

// Test 1: Session expired handling
it('should handle SESSION_EXPIRED error', () => {
  const mockError = {
    status: 401,
    error: { message: 'SESSION_EXPIRED' }
  } as HttpErrorResponse;

  // Verify interceptor dispatches sessionTimeoutAction
  // Verify sessionStorage is set with current URL
});

// Test 2: Non-session errors should pass through
it('should pass through non-401 errors', () => {
  const mockError = { status: 500 } as HttpErrorResponse;
  // Verify error is re-thrown
});
```

### Test Cases for Data Sanitization

The `hasFileAboveMaxSizeWarning()` method in `dynamic-datafiles.component.ts` uses `innerHTML` with a template string. While the template uses controlled values from app config, tests should verify:

```typescript
it('should sanitize warning message properly', () => {
  // Test with various config values
  // Verify no script tags or malicious content can be injected
});
```

**Affected Files**:
- `src/app/shared/interceptors/auth.interceptor.ts`: Lines 31-41 (error handling)
- `src/app/datasets/datafiles/dynamic-datafiles/dynamic-datafiles.component.ts`: Lines 438-466 (warning message generation)

---
---
## Summary

### Key Findings

1. **Feature Implementation**: The dynamic datafile table is a well-designed addition that provides configurable, scalable datafile viewing with server-side pagination and selection management.

2. **Architecture**: Follows existing patterns with proper NgRx state management, service integration, and component structure.

3. **Security**: No significant security vulnerabilities identified. The auth interceptor actually improves security posture.

4. **Testing Gaps**: 
   - **Critical**: Missing tests for `DynamicDatafilesComponent`
   - **High**: Missing tests for `AuthInterceptor`
   - **Medium**: Missing tests for `TimeDurationPipe` and enhanced coverage for state management

5. **Code Quality**:
   - Generally high quality with good separation of concerns
   - Some minor issues with unused imports and type casting
   - Documentation is comprehensive and updated

### Recommendations

1. **Before Merging**: Add tests for the three missing test files (DynamicDatafilesComponent, AuthInterceptor, TimeDurationPipe)
2. **Cleanup**: Remove unused imports in auth.interceptor.ts
3. **Type Safety**: Address type casting issues (`as unknown as FileOrigdatablock[]`) with proper type definitions
4. **Documentation**: The documentation updates are excellent - maintain this standard

### Overall Verdict

The changes are **APPROVED WITH CONDITIONS**. The feature is well-implemented and provides significant value. The main condition is adding comprehensive test coverage for the new components before merging to master.

---

Generated by Mistral Vibe.
Co-Authored-By: Mistral Vibe <vibe@mistral.ai>

**Comment by abdellah257** [comment]
Thanks @nitrosx for the review, I have put this branch on draft due to conflict with #2349, This effort preceded the submission of the PR mentioned, and If merged I intend to keep only the configuration part of this PR ( since we will be using only the dynamicTable ), and probably the file selection state management too.
